src/Core/Grid/Query/MerchandiseReturnProductsQueryBuilder.php

```diff
@@ -0,0 +1,202 @@
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Class MerchandiseReturnQueryBuilder builds queries for merchandise returns grid data.
+ */
+final class MerchandiseReturnProductsQueryBuilder extends AbstractDoctrineQueryBuilder
+{
+    /**
+     * @var int
+     */
+    private $contextLanguageId;
+
+    /**
+     * @var DoctrineSearchCriteriaApplicatorInterface
+     */
+    private $searchCriteriaApplicator;
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @param Connection $connection
+     * @param $dbPrefix
+     * @param DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+     * @param int $contextLanguageId
+     */
+    public function __construct(
+        Connection $connection,
+        $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator,
+        int $contextLanguageId,
+        /*
+         * @todo
+         * I need RequestStack to get merchandiseReturnId to query builder.
+         * I saw that categoryQueryBuilder uses filters for this purpose and it ends up using
+         * FilterCategorySearchCriteriaListener I wonder should I approach it the same way or stick to request stack?
+         * The filter way could be cleaner, but I still will use Request stack in DefinitionFactory so is it worth to
+         * do it via FilterSearchCriteriaListener here?
+         */
+        RequestStack $requestStack
+    ) {
+        parent::__construct($connection, $dbPrefix);
+        $this->searchCriteriaApplicator = $searchCriteriaApplicator;
+        $this->contextLanguageId = $contextLanguageId;
+        $this->requestStack = $requestStack;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getMerchandiseReturnProductsQueryBuilder($searchCriteria);
+        $qb
+            ->select('
+            ord.id_order_return,
+            ord.id_order_detail,
+            ord.product_quantity,
+            od.product_name,
+            od.product_reference,
+            od.id_customization,
+            o.id_cart'
+            )
+            ->groupBy('ord.id_order_detail');
+
+        $this->searchCriteriaApplicator
+            ->applySorting($searchCriteria, $qb)
+            ->applyPagination($searchCriteria, $qb);
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        return $this->getMerchandiseReturnProductsQueryBuilder($searchCriteria)
+            ->select('COUNT(ord.id_order_detail)');
+    }
+
+    /**
+     * @param SearchCriteriaInterface $searchCriteria
+     *
+     * @return QueryBuilder
+     */
+    private function getMerchandiseReturnProductsQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        if (null !== ($request = $this->requestStack->getCurrentRequest())
```

Comment:
You must initialize `$merchandiseReturnId` to some value before
```
$merchandiseReturnId = null
if (...)
```

It's a good practice and will be flagged as an issue by PHPStan when we enable it (and it's scheduled for PS 1.7.8)

Reply:
Moreover if `$merchandiseReturnId` is needed to perform the search, then you can simply throw an Exception if it cannot be retrieved from `$requestStack`.

"If class domain A.php cannot work without parameter X, it throws Exception"
It's up to application layer (= web controller, CLI command, or API controller) to handle this exception